core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java

```diff
@@ -148,41 +179,361 @@ public BigDecimal decimalValue() {
-				&& Objects.equals(getDatatype(), ((Literal) o).getDatatype());
+				&& getLabel().equals(((Literal) o).getLabel())
+				&& getDatatype().equals(((Literal) o).getDatatype())
+				&& equals(getLanguage(), ((Literal) o).getLanguage());
 	}
 
 	@Override
 	public int hashCode() {
-		return Objects.hashCode(getLabel());
+		return getLabel().hashCode();
 	}
 
-	/**
-	 * Returns the label of the literal with its language or datatype.
-	 * <p>
-	 * Note that this method does not escape the quoted label.
-	 */
 	@Override
 	public String toString() {
 
-		final String label = Optional.ofNullable(getLabel()).orElse("");
-		final String language = getLanguage().orElse(null);
-		final String datatype = Optional.ofNullable(getDatatype()).map(Value::stringValue).orElse(XSD_STRING);
+		final String label = '"' + getLabel() + '"';
+
+		return getLanguage()
+
+				.map(language -> label + '@' + language)
+
+				.orElseGet(() -> {
+
+					final IRI datatype = getDatatype();
+
+					return datatype.equals(XSD_STRING) ? label
+							: label + "^^<" + datatype.stringValue() + ">";
+
+				});
+	}
+
+	private boolean equals(Optional<String> x, Optional<String> y) {
+
+		final boolean px = x.isPresent();
+		final boolean py = y.isPresent();
+
+		return px && py && x.get().equalsIgnoreCase(y.get()) || !px && !py;
+	}
+
+	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	static class TypedLiteral extends AbstractLiteral {
+
+		private static final long serialVersionUID = -19640527584237291L;
+
+		private String label;
+		private IRI datatype;
+
+		TypedLiteral(String label) {
+			this.label = label;
+			this.datatype = XSD_STRING;
+		}
+
+		TypedLiteral(String label, IRI datatype) {
+			this.label = label;
+			this.datatype = (datatype != null) ? datatype : XSD_STRING;
+		}
+
+		@Override
+		public String getLabel() {
+			return label;
+		}
+
+		@Override
+		public Optional<String> getLanguage() {
+			return Optional.empty();
+		}
+
+		@Override
+		public IRI getDatatype() {
+			return datatype;
+		}
 
-		return language != null ? '"' + label + '"' + '@' + language
-				: datatype.equals(XSD_STRING) ? '"' + label + '"'
-						: '"' + label + '"' + "^^<" + datatype + ">";
 	}
 
-	private String normalize(String tag) {
-		return tag.toUpperCase(Locale.ROOT);
+	static class TaggedLiteral extends AbstractLiteral {
+
+		private static final long serialVersionUID = -19640527584237291L;
+
+		private String label;
+		private String language;
+
+		TaggedLiteral(String label, String language) {
+			this.label = label;
+			this.language = language;
+		}
+
+		@Override
+		public String getLabel() {
+			return label;
+		}
+
+		@Override
+		public Optional<String> getLanguage() {
+			return Optional.of(language);
+		}
+
+		@Override
+		public IRI getDatatype() {
+			return RDF_LANGSTRING;
+		}
+
+	}
+
+	static class BooleanLiteral extends AbstractLiteral {
```

Comment:
We already have a `BooleanLiteral` class. Do we need this?

Reply:
See #???